packages/expo-cli/src/accounts.ts

```diff
@@ -95,7 +110,165 @@ export async function login(options: CommandOptions): Promise<User> {
+    device => device.method === UserSecondFactorDeviceMethod.SMS
+  );
+
+  const authenticatorChoiceSentinel = -1;
+  const cancelChoiceSentinel = -2;
+
+  const deviceChoices = smsNonPrimarySecondFactorDevices.map((device, idx) => ({
+    name: device.sms_phone_number!,
+    value: idx,
+  }));
+
+  if (hasAuthenticatorSecondFactorDevice) {
+    deviceChoices.push({
+      name: 'Authenticator',
+      value: authenticatorChoiceSentinel,
+    });
+  }
+
+  deviceChoices.push({
+    name: 'Cancel',
+    value: cancelChoiceSentinel,
+  });
+
+  const question: Question = {
+    type: 'list',
+    name: 'choice',
+    message: 'Select a second-factor device:',
+    choices: deviceChoices,
+    pageSize: Infinity,
+  };
+
+  const { choice } = await prompt(question);
+  if (choice === cancelChoiceSentinel) {
+    throw new Error('Cancelled login');
+  } else if (choice === authenticatorChoiceSentinel) {
+    return await _promptForOTPAsync('cancel');
+  }
+
+  const device = smsNonPrimarySecondFactorDevices[choice];
+
+  const apiAnonymous = ApiV2.clientForUser();
+  await apiAnonymous.postAsync('auth/send-sms-otp', {
+    username,
+    password,
+    secondFactorDeviceID: device.id,
+  });
+
+  return await _promptForOTPAsync('cancel');
+}
+
+/**
+ * Handle the special case error indicating that a second-factor is required for
+ * authentication.
+ *
+ * There are three cases we need to handle:
+ * 1. User's primary second-factor device was SMS, OTP was automatically sent by the server to that
+ *    device already. In this case we should just prompt for the SMS OTP (or backup code), which the
+ *    user should be receiving shortly. We should give the user a way to cancel and the prompt and move
+ *    to case 3 below.
+ * 2. User's primary second-factor device is authenticator. In this case we should prompt for authenticator
+ *    OTP (or backup code) and also give the user a way to cancel and move to case 3 below.
+ * 3. User doesn't have a primary device or doesn't have access to their primary device. In this case
+ *    we should show a picker of the SMS devices that they can have an OTP code sent to, and when
+ *    the user picks one we show a prompt for the sent OTP.
+ */
+async function _retryUsernamePasswordAuthWithOTPAsync(
+  username: string,
+  password: string,
+  metadata: {
+    secondFactorDevices?: SecondFactorDevice[];
+    smsAutomaticallySent?: boolean;
+  }
+): Promise<User> {
+  const { secondFactorDevices, smsAutomaticallySent } = metadata;
+  invariant(
+    secondFactorDevices !== undefined && smsAutomaticallySent !== undefined,
+    `Malformed OTP error metadata: ${metadata}`
+  );
+
+  const primaryDevice = secondFactorDevices.find(device => device.is_primary);
+  let otp: string | null = null;
+
+  if (smsAutomaticallySent) {
+    invariant(
+      primaryDevice,
+      'OTP should only automatically be sent when there is a primary device'
```

Comment:
Since we know the cause of the problem, perhaps we could format this more like a warning rather than an error. For example, when someone runs `expo start --config` instead of throwing an invalid type, we warn how to use flag and exit with non-zero. https://github.com/expo/expo-cli/blob/50f98f66a52cb3371aa580d9d504d465ce8558a6/packages/expo-cli/src/exp.ts#L418-L422 the lack of generic error and a stack trace helps people know that CLI isn't buggy.

Reply:
This is also a constraint that is easy to express on the server but hard to express over an API. Essentially, when an SMS is automatically sent that means that there is a primary device and it is an SMS device. The invariant here is to appease TypeScript to ensure that this is non-null for use.